invokeai/app/services/shared/invocation_context.py

```diff
@@ -0,0 +1,625 @@
+
+            # If the invocation inherits metadata, use that. Else, use the metadata passed in.
+            metadata_ = (
+                context_data.invocation.metadata if isinstance(context_data.invocation, WithMetadata) else metadata
+            )
+
+            return services.images.create(
+                image=image,
+                is_intermediate=context_data.invocation.is_intermediate,
+                image_category=image_category,
+                board_id=board_id,
+                metadata=metadata_,
+                image_origin=ResourceOrigin.INTERNAL,
+                workflow=context_data.workflow,
+                session_id=context_data.session_id,
+                node_id=context_data.invocation.id,
+            )
+
+        def get_pil(image_name: str) -> Image:
+            """
+            Gets an image as a PIL Image object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_pil_image(image_name)
+
+        def get_metadata(image_name: str) -> Optional[MetadataField]:
+            """
+            Gets an image's metadata, if it has any.
+
+            :param image_name: The name of the image to get the metadata for.
+            """
+            return services.images.get_metadata(image_name)
+
+        def get_dto(image_name: str) -> ImageDTO:
+            """
+            Gets an image as an ImageDTO object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_dto(image_name)
+
+        def update(
+            image_name: str,
+            board_id: Optional[str] = None,
+            is_intermediate: Optional[bool] = False,
+        ) -> ImageDTO:
+            """
+            Updates an image, returning its updated DTO.
+
+            It is not suggested to update images saved by earlier nodes, as this can cause confusion for users.
+
+            If you use this method, you *must* return the image as an :class:`ImageOutput` for the gallery to
+            get the updated image.
+
+            :param image_name: The name of the image to update.
+            :param board_id: The board ID to add the image to, if it should be added.
+            :param is_intermediate: Whether the image is an intermediate. Intermediate images aren't added to the gallery.
+            """
+            if is_intermediate is not None:
+                services.images.update(image_name, ImageRecordChanges(is_intermediate=is_intermediate))
+            if board_id is None:
+                services.board_images.remove_image_from_board(image_name)
+            else:
+                services.board_images.add_image_to_board(image_name, board_id)
+            return services.images.get_dto(image_name)
+
+        self.save = save
+        self.get_pil = get_pil
+        self.get_metadata = get_metadata
+        self.get_dto = get_dto
+        self.update = update
+
+
+class LatentsInterface:
```

Comment:
Sure.

Reply:
Ok, I've removed all (? think I got em all) references to "latents" within the services. See my latest commit on the whole PR for detail.

I don't want to change the invocation definitions or output classes though, because that would a massively breaking change.